Tests/SwiftFusionTests/Inference/Scratch.swift

```diff
@@ -0,0 +1,173 @@
+import Foundation
+import TensorFlow
+import XCTest
+
+import PenguinStructures
+@testable import SwiftFusion
+
+/// A factor on two discrete labels evaluation the transition probability
+struct DiscreteTransitionFactor {
+  /// The number of states.
+  let stateCount: Int
+
+  /// Entry `i * stateCount + j` is the probability of transitioning from state `j` to state `i`.
+  let transitionMatrix: [Double]
+
+  init(
+    _ inputId1: TypedID<Int, Int>,
+    _ inputId2: TypedID<Int, Int>,
+    _ stateCount: Int,
+    _ transitionMatrix: [Double]
+  ) {
+    precondition(transitionMatrix.count == stateCount * stateCount)
+    self.stateCount = stateCount
+    self.transitionMatrix = transitionMatrix
+  }
+
+  func error(at label1: Int, _ label2: Int) -> Double {
+    return -log(transitionMatrix[label2 * stateCount + label1])
+  }
+}
+
+/// A factor with a switchable motion model.
+///
+/// `JacobianRows` specifies the `Rows` parameter of the Jacobian of this factor. See the
+/// documentation on `NewJacobianFactor.jacobian` for more information. Use the typealiases below to
+/// avoid specifying this type parameter every time you create an instance.
+public struct NewSwitchingBetweenFactor<Pose: LieGroup, JacobianRows: FixedSizeArray>:
+  NewLinearizableFactor
+  where JacobianRows.Element == Tuple2<Pose.TangentVector, Pose.TangentVector>
+{
+  public typealias Variables = Tuple3<Pose, Int, Pose>
+
+  public let edges: Variables.Indices
+  
+  /// Movement temmplates for each label.
+  let movements: [Pose]
+
+  public init(_ from: TypedID<Pose, Int>,
+              _ label: TypedID<Int, Int>,
+              _ to: TypedID<Pose, Int>,
+              _ movements: [Pose]) {
+    self.edges = Tuple3(from, label, to)
+    self.movements = movements
+  }
+
+  public typealias ErrorVector = Pose.TangentVector
+
+  @differentiable(wrt: (start, end))
+  public func errorVector(_ start: Pose, _ label: Int, _ end: Pose) -> ErrorVector {
+    let actualMotion = between(start, end)
+    return movements[label].localCoordinate(actualMotion)
+  }
+
+  // Note: All the remaining code in this factor is boilerplate that we can eventually eliminate
+  // with sugar.
+  
+  public func error(at x: Variables) -> Double {
+    return errorVector(at: x).squaredNorm
+  }
+
+  public func errorVector(at x: Variables) -> Pose.TangentVector {
+    return errorVector(x.head, x.tail.head, x.tail.tail.head)
+  }
+
+  public typealias Linearization = NewJacobianFactor<JacobianRows, ErrorVector>
+  public func linearized(at x: Variables) -> Linearization {
+    Linearization(linearizing: errorVector, at: x, edges: edges)
+  }
```

Comment:
The linearization of this factor has 2 inputs because the discrete input to the nonlinear factor has no corresponding input in the linearization.

So you need to help `init(linearizing:at:edges)` figure out what the relevant inputs are:
* Partially apply `errorVector` to the discrete input before passing it to `linearizing:`
* Construct the 2-tuple of differentiable variable ids, and pass it to `edges:`

I'll reply with some working code in a moment.

Reply:
Here is the code:
```
  public func linearized(at x: Variables) -> Linearization {
    let (start, label, end) = (x.head, x.tail.head, x.tail.tail.head)
    let (startEdge, endEdge) = (edges.head, edges.tail.tail.head)
    let differentiableVariables = Tuple2(start, end)
    let differentiableEdges = Tuple2(startEdge, endEdge)
    return Linearization(
      linearizing: { differentiableVariables in
        let (start, end) = (differentiableVariables.head, differentiableVariables.tail.head)
        return errorVector(start, label, end)
      },
      at: differentiableVariables,
      edges: differentiableEdges
    )
  }
```

Reply:
Nothing very interesting is happening inside that `linearized` method. I'll try to eliminate it in my boilerplate-elimination work this week.